enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,356 @@
+persistent volumes to disable or switch those technologies and reduce the
+energy, memory, CPU, disk/over-the-wire and boot time footprint.
+
+In addition, upstream versions of MicroShift may need to ship with different CNI/CSI
+options or image versions, and we are looking for ways to enable this.
+
+This same implementation could potentially be applied to other components in
+future enhancements.
+
+### User Stories
+
+As a MicroShift Deployment admin persona, I want to disable the CSI plugin since
+my device has limited capacity and my application does not need PVs but would
+benefit from the additional RAM and CPU.
+
+As a MicroShift Deployment admin persona, I want to disable default CSI and CNI
+plugins on my device to save RAM, CPU and disk footprint as my application
+does not benefit from those features.
+
+As an Upstream Developer, I need to run a different set of public available images
+and potentially other configurations for the CSI/CNI components, I can release new
+packages for those components to be tested and used upstream.
+
+As an Project maintainer I want to avoid increasing coupling between the MicroShift
+core and the CSI/CNI or other plugins.
+
+As a Project mantainer I want to allow other teams to create new components
+which can be installed to MicroShift without any need of coupling to the MicroShift
+project or repository.
+
+### Goals
+* Allow networking and storage core components to be optionally installed.
+
+* Allow networking and storage components to be replaced upstream for upstream
+  versions of those components.
+
+* Split the RPM packages to resemble those optional components.
+
+* Systemd services of the then optional networking components (like `microshift-ovs-init`)
+  should still start before the `microshift` systemd service, but `microshift` cannot
+  depend on `microshift-ovs-init` anymore. This can be acomplished by using the
+  Before=microshift clause in `microshift-ovs-init.service` and enabling this service
+  during the `microshift-networking-ovnk` sub-package install.
+
+* The rebase script should handle extracting and updating the component asset yamls
+  into the new directory, generating the right kustomize / metadata yamls.
+
+* Factor out all CNI/CSI specific code from MicroShift code, so existing and new plugins
+  can be developed and mantained independently from MicroShift.
+
+* As much as possible, use mechanisms familiar to K8s users (for manifests) or Linux
+  users (for host-config/RPMs), rather than inventing new mechanisms.
+
+### Non-Goals
+
+* Making DNS component optional.
+
+* Making the certificate management component optional.
+
+* Allowing the replacement of network or storage components on runtime. Once
+  the system has booted to a storage or network component, the component cannot
+  be switched (but a full reset of the system `cleanup-all-microshift-data` and restart
+  should work for developers).
+
+## Proposal
```

Comment:
Do we want to split the plugin code/yamls to separate repo or is it kept in the same microshift main repo? I assume it is latter since we don't mention the repo split.